doc/man/tdx-util.8
.\" Automatically generated by Pod::Man 4.07 (Pod::Simple 3.32)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.if !\nF .nr F 0
.if \nF>0 \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    if !\nF==2 \{\
.        nr % 0
.        nr F 2
.    \}
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "TDX-UTIL 8"
.TH TDX-UTIL 8 "2015-09-12" "INN 2.6.3" "InterNetNews Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
tdx\-util \- Tradindexed overview manipulation utility
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBtdx-util\fR [\fB\-AFcgiOo\fR] [\fB\-a\fR \fIarticle\fR] [\fB\-f\fR \fIstatus\fR]
[\fB\-n\fR \fInewsgroup\fR] [\fB\-p\fR \fIpath\fR] [\fB\-R\fR \fIpath\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBtdx-util\fR is an administrative interface to the tradindexed overview
method for \s-1INN. \s0 It only works on tradindexed overview databases, not on
any other type of \s-1INN\s0 overview.  It allows the administrator to dump
various information about the internal state of the overview, audit it for
errors, and rebuild portions of the overview database.
.PP
The tradindexed overview method should lock properly and therefore it
should be safe to run this utility and perform any operation it performs,
including full repairs or per-group overview rebuilds, while the server is
running.  However, note that some of the operations performed by this
utility can take an extended period of time and will hold locks in the
overview database during that period, which depending on what the server
is doing may cause the server to stall until \fBtdx-util\fR completes its
operation.
.PP
The dump operations are \fB\-i\fR, which dumps the master index for the
overview database, \fB\-g\fR, which dumps the index for an individual group,
and \fB\-o\fR and \fB\-O\fR, which dump the overview information for a particular
group (including the additional metadata stored in the index) in two
different formats.  For \fB\-g\fR, \fB\-o\fR, and \fB\-O\fR, the \fB\-n\fR option must
also be given to specify a newsgroup to operate on.
.PP
To add a new newsgroup to the overview database, use \fB\-c\fR.  A group must
be specified with \fB\-n\fR.  If the group status is something other than \f(CW\*(C`y\*(C'\fR,
it should be specified with \fB\-f\fR, and the low and high article numbers
may be specified with \fB\-a\fR.  If only one number is given rather than a
range, it will be taken to be the high water mark and the low mark will be
set to 1.
.PP
To audit the entire overview database for problems, use \fB\-A\fR.  Any
problems found will be reported to standard error.  Use \fB\-F\fR to correct
the errors found.
.PP
To rebuild the database for a particular newsgroup, use \fB\-R\fR.  The \fB\-R\fR
option takes a path to a directory which contains all of the articles for
that newsgroup, one per file.  The names of the files must be the numbers
of the articles in that group.  (In other words, this directory must be a
traditional spool directory for that group.)  The \fB\-n\fR option must also
be given to specify the newsgroup for which the overview is being rebuilt.
.PP
For all operations performed by \fBtdx-util\fR, a different overview database
than the one specified in \fIinn.conf\fR may be specified using the \fB\-p\fR
option.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-A\fR" 4
.IX Item "-A"
Audit the entire overview database for problems.  This runs the internal
consistency checks built into the tradindexed overview implementation,
checking such things as the validity and reachability of all group index
entries, the format of the individual overview entries, the correspondence
of index entries to overview data, and similar such things.  No changes
will be made to the database, but problems will be reported to standard
error.
.IP "\fB\-a\fR \fIarticle\fR" 4
.IX Item "-a article"
The article number or numbers to act on.  \fIarticle\fR is a valid \s-1NNTP\s0
range, meaning that it can be either a single article number or a range of
article numbers (like \f(CW\*(C`1\-5\*(C'\fR).  Either the start or the end (or both) of
the range may be omitted, in which case they will be set to the first or
last article number in the group.  Passing \f(CW\*(C`\-\*(C'\fR for \fIarticle\fR is
therefore equivalent to not using the \fB\-a\fR option at all.
.Sp
Only useful in combination with the \fB\-o\fR option to dump overview
information or with \fB\-c\fR to specify the low and high article numbers when
creating a group.
.IP "\fB\-c\fR" 4
.IX Item "-c"
Create a new group in the overview database.  The group must be specified
with \fB\-n\fR.  The newsgroup status defaults to \f(CW\*(C`y\*(C'\fR but may be set with
\&\fB\-f\fR.  The low and high article numbers default to 1 and 0 respectively,
but may be set with \fB\-a\fR.  If only one number is given to \fB\-a\fR, it is
taken as the high article number.
.IP "\fB\-F\fR" 4
.IX Item "-F"
Audit the entire overview database for problems, fixing them as they're
detected where possible.  This runs the internal consistency checks built
into the tradindexed overview implementation, checking such things as the
validity and reachability of all group index entries, the format of the
individual overview entries, the correspondence of index entries to
overview data, and similar such things.  The strategy used when fixing
problems is to throw away data that's unrecoverable, so be warned that
using this option may result in inaccessible articles if their overview
data has been corrupted.
.Sp
To see what would be changed by \fB\-F\fR, run \fBtdx-util\fR with \fB\-A\fR first.
.IP "\fB\-f\fR \fIstatus\fR" 4
.IX Item "-f status"
When creating a newsgroup with \fB\-c\fR, set the status of the newly created
group to \fIstatus\fR instead of \f(CW\*(C`y\*(C'\fR.  Only useful with \fB\-c\fR.
.IP "\fB\-g\fR" 4
.IX Item "-g"
Dump the index of a particular group.  The fields are, in order, the
article number, the offset of the data for that article in the overview
data file for that group, the length of the overview data, the time (in
seconds since epoch) when the article arrived on the server, the time (in
seconds since epoch) when the article should expire based on its Expires:
header (or 0 if there is no Expires: header), and the storage \s-1API\s0 token of
the article.
.Sp
If this option is given, the \fB\-n\fR option must also be given to specify
the newsgroup on which to act.
.IP "\fB\-i\fR" 4
.IX Item "-i"
Dump the master index of the overview database.  This contains similar
information to the server \fIactive\fR file, such as high and low water marks
and moderation status, and is the information that \fBnnrpd\fR hands out to
clients.
.Sp
The fields are, in order, the newsgroup name, the high water mark, the low
water mark, the base article number (the point at which the group index
begins), the count of articles in the group, the group status, the
time (in seconds since epoch) when that newsgroup was deleted or 0 if it
hasn't been, and the inode of the index file for that group.
.Sp
A particular newsgroup can be specified with the \fB\-n\fR option.  If \fB\-n\fR
is not given, the entire master index will be dumped.
.IP "\fB\-n\fR \fInewsgroup\fR" 4
.IX Item "-n newsgroup"
Specify the newsgroup on which to act, required for the \fB\-i\fR, \fB\-o\fR, and
\&\fB\-R\fR options.
.IP "\fB\-O\fR" 4
.IX Item "-O"
Dump the overview information for a newsgroup in the format used by
\&\fBoverchan\fR as input.  Each line will start with the storage \s-1API\s0 token,
the arrival timestamp in seconds since epoch, the expires timestamp in the
same format (or 0 if there is no Expires: header), and then the overview
data.
.Sp
If this option is given, the \fB\-n\fR option must also be given to specify
the newsgroup on which to act.  By default, all of the overview
information for that newsgroup is dumped, but the \fB\-a\fR option may be
given to restrict the dump to the information for a single article.
.IP "\fB\-o\fR" 4
.IX Item "-o"
Dump the overview information for a newsgroup, in the same format as it
would be returned to clients but with one modification.  Appended to the
end of each entry will be four additional pieces of data:  the article
number according to the index file for that group labelled with
\&\f(CW\*(C`Article:\*(C'\fR, the storage \s-1API\s0 token for that article labelled with
\&\f(CW\*(C`Token:\*(C'\fR, the arrival date for that article on the server in \s-1RFC\s0\ 5322 date
format labelled with \f(CW\*(C`Arrived:\*(C'\fR, and the expiration date for that article
(from the Expires: header) in \s-1RFC\s0\ 5322 date format if there is any, labelled
with \f(CW\*(C`Expires:\*(C'\fR.
.Sp
If this option is given, the \fB\-n\fR option must also be given to specify
the newsgroup on which to act.  By default, all of the overview
information for that newsgroup is dumped, but the \fB\-a\fR option may be
given to restrict the dump to the information for a single article.
.IP "\fB\-p\fR \fIpath\fR" 4
.IX Item "-p path"
Act on the overview database rooted in \fIpath\fR, overriding the overview
path specified in \fIinn.conf\fR.
.IP "\fB\-R\fR \fIpath\fR" 4
.IX Item "-R path"
Rebuild the overview for a given group from the articles stored in
\&\fIpath\fR.  The articles must be in the form of a traditional spool
directory; in other words, each article must be in a separate file and the
name of the file must match the article number of the article.
.Sp
If this option is given, the \fB\-n\fR option must also be given to specify
the newsgroup on which to act.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
Dump the master index for the overview database in \fI/news/overview\fR,
regardless of the overview path specified in \fIinn.conf\fR:
.PP
.Vb 1
\&    tdx\-util \-i \-p /news/overview
.Ve
.PP
Dump the group index for example.test:
.PP
.Vb 1
\&    tdx\-util \-g \-n example.test
.Ve
.PP
Dump the complete overview information for example.test:
.PP
.Vb 1
\&    tdx\-util \-o \-n example.test
.Ve
.PP
Dump the overview information for articles 45 and higher in example.test:
.PP
.Vb 1
\&    tdx\-util \-o \-n example.test \-a 45\-
.Ve
.PP
Create an entry for example.test with mode m and low and high article
numbers of 4 and 23, respectively.
.PP
.Vb 1
\&    tdx\-util \-c \-n example.test \-f m \-a 4\-23
.Ve
.PP
Audit the entire overview database for any problems:
.PP
.Vb 1
\&    tdx\-util \-A
.Ve
.PP
Rebuild the overview information for example.test from a traditional spool
directory:
.PP
.Vb 1
\&    tdx\-util \-R <patharticles in inn.conf>/example/test \-n example.test
.Ve
.PP
The last command may be useful for recovering from a bad crash or
corrupted overview information for a particular group, if you are also
using the tradspool article storage method.
.SH "HISTORY"
.IX Header "HISTORY"
Written by Russ Allbery <eagle@eyrie.org> for InterNetNews.
.PP
\&\f(CW$Id:\fR tdx\-util.pod 9767 2014\-12\-07 21:13:43Z iulius $
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fImakehistory\fR\|(8), \fInnrpd\fR\|(8).


NEWS
Changes in 2.6.3

    * Fixed the selection of the elliptic curve to use with OpenSSL 1.1.0 or
      later; NIST P-256 was enforced instead of using the most secure curve.

    * A new inn.conf parameter has been added to fine-tune the cipher suites
      to use with TLS 1.3:  the *tlsciphers13* now permits configuring them.
      A separate cipher suite configuration parameter is needed for TLS 1.3
      because TLS 1.3 cipher suites are not compatible with TLS 1.2, and
      vice-versa.  In order to avoid issues where legacy TLS 1.2 cipher
      suite configuration configured in the *tlsciphers* parameter would
      inadvertently disable all TLS 1.3 cipher suites, the inn.conf
      configuration has been separated out.

    * Fixed a regression since INN 2.6.1 that prevented articles with
      internationalized header fields (that is to say encoded in UTF-8) from
      being posted.

    * Support for Python 3 has been added to INN.  Embedded Python filtering
      and authentication hooks for innd and nnrpd can now use version 3.3.0
      or later of the Python interpreter.  In the 2.x series, version 2.3.0
      or later is still supported.

      When configuring INN with the --with-python flag, the "PYTHON"
      environment variable, when set, is used to select the interpreter to
      embed.  Otherwise, it is searched in standard paths.

      In case you change the Python interpreter to embed, make sure that the
      Python scripts you use are written in the expected syntax for that
      version of the Python interpreter.  Notably, buffer objects have been
      replaced with memoryview objects in Python 3, and UTF-8 encoding now
      really matters for string literals (Python 3 uses bytes and Unicode
      objects).

      INN documentation and samples of Python hooks have been updated to
      provide more examples.

    * When a Python or Perl filter hook rejects an article, innd now
      mentions the reason in response to CHECK and TAKETHIS commands. 
      Previously, the reason was given only for the IHAVE command.

    * nnrpd now properly logs the hostname of clients whose connection
      failed owing to an issue during the negotiation of a TLS session or
      high load average.

Changes in 2.6.2

    * A new *syntaxchecks* parameter has been added in inn.conf.  It permits
      controlling the level of checks performed by innd and nnrpd.  Up to
      now, only one check can be enabled/disabled:  when *laxmid* is
      mentioned in the values of this new parameter, INN accepts Message-IDs
      that contain ".." in the left part, as well as Message-IDs with two
      "@" (such Message-IDs would otherwise be considered as syntactically
      invalid).  See the inn.conf(5) man page for more details.

      The check is disabled by default (*no-laxmid*), which corresponds to
      the legacy behaviour of INN 2.6.1 and earlier.

    * Use of the ovdb_server helper server is now the default when using the
      ovdb overview method, that is to say the default value for the
      *readserver* parameter in ovdb.conf is now set to true.  It improves
      stability and avoids deadlocks, timing issues and corrupted ovdb
      databases.

    * mailpost now removes empty header fields before attempting to post
      articles, and keeps trace of them in the X-Mailpost-Empty-Hdrs: newly
      generated header field body.  Also, mailpost now sanitizes header
      fields with regards to empty continuation header lines.  Thanks to
      Kamil Jonca for these bug reports.

    * A new -z parameter has been added to mailpost to mention a list of
      header fields to remove from the gated message.  Thanks to Dieter
      Stussy for the patch.

    * Fixed a bug in inews that was rejecting articles containing header
      fields whose length exceeded 998 bytes.  This limitation is for the
      length of a single line of a header field (and not for the length of
      the whole header field, as it was wrongly the case).

    * Added support for GnuPG's gpg binary (in addition to gpgv) in
      pgpverify.  Indeed, gpg still validates signatures made with weak
      digest algorithms like MD5 whereas gpgv no longer do.  Thanks to
      Thomas Hochstein for the patch, which permits validating control
      articles for hierarchies that are still using old PGP keys.

    * Added similar support for GnuPG's gpg binary in perl-nocem to validate
      NoCeM notices from issuers who are still using old PGP keys.

    * A few commands listed in the "Control commands to INND" section in
      daily Usenet reports were appearing as a mere letter; all of them are
      now properly converted to meaningful words.

    * The *tlsprotocols* parameter in inn.conf now recognizes the "TLSv1.3"
      value (for OpenSSL versions implementing TLS 1.3, that is to say
      starting from OpenSSL 1.1.1).

    * The buffindexed overview method will now hopefully work properly on
      systems with a native page size larger than 16KB.

    * Other minor bug fixes and documentation improvements.

Changes in 2.6.1

    * nnrpd now uses -0000 as the time zone for Date: and Injection-Date:
      header fields it generates.  It was previously using +0000, wrongly
      systematically indicating a local time zone at Universal Time when
      *localtime* is set to false (which is the default) in readers.conf. 
      The +0000 time zone will now be used only if *localtime* is set to
      true and UTC is really the local time zone of the server.

    * Julien Elie has implemented in nnrpd the new COMPRESS command
      described in draft-murchison-nntp-compress that extends the NNTP
      protocol to allow a connection to be effectively and efficiently
      compressed.  News clients that also support that extension will be
      able to benefit from that bandwidth optimization and improvement in
      speed.  Moreover, using COMPRESS is more secure than TLS-level

storage/tradindexed/tdx-util.c
/*  $Id: tdx-util.c 9019 2010-03-19 21:27:15Z iulius $
**
**  Utility for managing a tradindexed overview spool.
**
**  This utility can manipulate a tradindexed overview spool in various ways,
**  including some ways that are useful for recovery from crashes.  It allows
**  the user to view the contents of the various data structures that
**  tradindexed stores on disk.
*/

#include "config.h"
#include "clibrary.h"
#include <ctype.h>
#include <dirent.h>
#include <sys/stat.h>

#include "inn/buffer.h"
#include "inn/libinn.h"
#include "inn/history.h"
#include "inn/innconf.h"
#include "inn/messages.h"
#include "inn/newsuser.h"
#include "inn/ov.h"
#include "inn/paths.h"
#include "inn/vector.h"
#include "inn/wire.h"
#include "ovinterface.h"
#include "tdx-private.h"
#include "tdx-structure.h"

/*
**  Dump the main index data, either all of it or that for a particular group
**  if the group argument is non-NULL.
*/
static void
dump_index(const char *group)
{
    struct group_index *index;

    index = tdx_index_open(OV_READ);
    if (index == NULL)
        return;
    if (group == NULL)
        tdx_index_dump(index, stdout);
    else {
        const struct group_entry *entry;

        entry = tdx_index_entry(index, group);
        if (entry == NULL) {
            warn("cannot find group %s", group);
            return;
        }
        tdx_index_print(group, entry, stdout);
    }
    tdx_index_close(index);
}


/*
**  Dump the data index file for a particular group.
*/
static void
dump_group_index(const char *group)
{
    struct group_index *index;
    struct group_entry *entry;
    struct group_data *data;

    index = tdx_index_open(OV_READ);
    if (index == NULL)
        return;
    entry = tdx_index_entry(index, group);
    if (entry == NULL) {
        warn("cannot find group %s in the index", group);
        return;
    }
    data = tdx_data_open(index, group, entry);
    if (data == NULL) {
        warn("cannot open group %s", group);
        return;
    }
    tdx_data_index_dump(data, stdout);
    tdx_data_close(data);
    tdx_index_close(index);
}


/*
**  Dump the overview data for a particular group.  If number is 0, dump the
**  overview data for all current articles; otherwise, only dump the data for
**  that particular article.  Include the article number, token, arrived time,
**  and expires time (if any) in the overview data as additional fields
**  normally.  If overchan is true, instead write out the data in the format
**  that overchan would expect as input.
*/
static void
dump_overview(const char *group, ARTNUM low, ARTNUM high, bool overchan)
{
    struct group_index *index;
    struct group_data *data;
    struct group_entry *entry;
    struct article article;
    struct search *search;
    char *p;
    char datestring[256];

    index = tdx_index_open(OV_READ);
    if (index == NULL)
        return;
    entry = tdx_index_entry(index, group);
    if (entry == NULL) {
        warn("cannot find group %s", group);
        return;
    }
    data = tdx_data_open(index, group, entry);
    if (data == NULL) {
        warn("cannot open group %s", group);
        return;
    }
    data->refcount++;

    if (low == 0)
        low = entry->low;
    if (high == 0)
        high = entry->high;
    search = tdx_search_open(data, low, high, entry->high);

    if (search == NULL) {
        if (low == high)
            puts("Article not found");
        else
            warn("cannot open search in %s: %lu - %lu", group, low, high);
        return;
    }
    while (tdx_search(search, &article)) {
        if (overchan) {
            p = memchr(article.overview, '\t', article.overlen - 3);
            if (p == NULL)
                continue;
            p++;
            printf("%s %lu %lu ", TokenToText(article.token),
                   (unsigned long) article.arrived,
                   (unsigned long) article.expires);
            fwrite(p, article.overlen - 2 - (p - article.overview), 1, stdout);
        } else {
            fwrite(article.overview, article.overlen - 2, 1, stdout);
            printf("\tArticle: %lu\tToken: %s", article.number,
                   TokenToText(article.token));
            makedate(article.arrived, true, datestring, sizeof(datestring));
            printf("\tArrived: %s", datestring);
            if (article.expires != 0) {
                makedate(article.expires, true, datestring,
                         sizeof(datestring));
                printf("\tExpires: %s", datestring);
            }
        }
        printf("\n");
    }
    tdx_search_close(search);
    tdx_data_close(data);
    tdx_index_close(index);
}


/*
**  Check a string to see if its a valid number.
*/
static bool
check_number(const char *string)
{
    const char *p;

    for (p = string; *p != '\0'; p++)
        if (!isdigit((unsigned char) *p))
            return false;
    return true;
}


/*
**  Find the message ID in the group overview data and return a copy of it.
**  Caller is responsible for freeing.
*/
static char *
extract_messageid(const char *overview)
{
    const char *p, *end;
    int count;

    for (p = overview, count = 0; count < 4; count++) {
        p = strchr(p + 1, '\t');
        if (p == NULL)
            return NULL;
    }
    p++;
    end = strchr(p, '\t');
    if (end == NULL)
        return NULL;
    return xstrndup(p, end - p);
}


/*
**  Compare two file names assuming they're numbers, used to sort the list of
**  articles numerically.  Suitable for use as a comparison function for
**  qsort.
*/
static int
file_compare(const void *p1, const void *p2)
{
    const char *file1 = *((const char * const *) p1);
    const char *file2 = *((const char * const *) p2);
    ARTNUM n1, n2;

    n1 = strtoul(file1, NULL, 10);
    n2 = strtoul(file2, NULL, 10);
    if (n1 > n2)
        return 1;
    else if (n1 < n2)
        return -1;
    else
        return 0;
}


/*
**  Get a list of articles in a directory, sorted by article number.
*/
static struct vector *
article_list(const char *directory)
{
    DIR *articles;
    struct dirent *file;
    struct vector *list;

    list = vector_new();
    articles = opendir(directory);
    if (articles == NULL)
        sysdie("cannot open directory %s", directory);
    while ((file = readdir(articles)) != NULL) {
        if (!check_number(file->d_name))
            continue;
        vector_add(list, file->d_name);
    }
    closedir(articles);

    qsort(list->strings, list->count, sizeof(list->strings[0]), file_compare);
    return list;
}


/*
**  Create a new newsgroup in the overview database.  Takes the group flag and
**  the high and low water marks.
*/
static void
group_create(const char *group, ARTNUM low, ARTNUM high, char flag)
{
    struct group_index *index;

    index = tdx_index_open(true);
    if (index == NULL)
        die("cannot open group index");
    if (low == 0)
        low = 1;
    if (!tdx_index_add(index, group, low, high, &flag))
        die("cannot create group %s", group);
    tdx_index_close(index);
}


/*
**  Rebuild the overview data for a particular group.  Takes a path to a
**  directory containing all the articles, as individual files, that should be
**  in that group.  The names of the files should be the article numbers in
**  the group.
*/
static void
group_rebuild(const char *group, const char *path)
{
    char *filename, *histpath, *article, *wireformat, *p;
    size_t size, length, file;
    int flags;
    struct buffer *overview = NULL;
    struct vector *extra, *files;
    struct history *history;
    struct group_index *index;
    struct group_data *data;
    struct group_entry *entry, info;
    struct article artdata;
    struct stat st;

    index = tdx_index_open(OV_READ);
    if (index == NULL)
        die("cannot open group index");
    entry = tdx_index_entry(index, group);
    if (entry == NULL) {
        if (!tdx_index_add(index, group, 1, 0, NF_FLAG_OK_STRING))
            die("cannot create group %s", group);
        entry = tdx_index_entry(index, group);
        if (entry == NULL)
            die("cannot find group %s", group);
    }
    info = *entry;
    data = tdx_data_rebuild_start(group);
    if (data == NULL)
        die("cannot start data rebuild for %s", group);
    if (!tdx_index_rebuild_start(index, entry))
        die("cannot start index rebuild for %s", group);

    histpath = concatpath(innconf->pathdb, INN_PATH_HISTORY);
    flags = HIS_RDONLY | HIS_ONDISK;
    history = HISopen(histpath, innconf->hismethod, flags);
    if (history == NULL)
        sysdie("cannot open history %s", histpath);
    free(histpath);

    extra = overview_extra_fields(true);
    files = article_list(path);

    info.count = 0;
    info.high = 0;
    info.low = 0;
    for (file = 0; file < files->count; file++) {
        filename = concatpath(path, files->strings[file]);
        article = ReadInFile(filename, &st);
        size = st.st_size;
        if (article == NULL) {
            syswarn("cannot read in %s", filename);
            free(filename);
            continue;
        }

        /* Check to see if the article is not in wire format.  If it isn't,
           convert it.  We only check the first line ending. */
        p = strchr(article, '\n');
        if (p != NULL && (p == article || p[-1] != '\r')) {
            wireformat = wire_from_native(article, size, &length);
            free(article);
            article = wireformat;
            size = length;
        }

        artdata.number = strtoul(files->strings[file], NULL, 10);
        if (artdata.number > info.high)
            info.high = artdata.number;
        if (artdata.number < info.low || info.low == 0)
            info.low = artdata.number;
        info.count++;
        overview = overview_build(artdata.number, article, size, extra,
                                  overview);
        artdata.overview = overview->data;
        artdata.overlen = overview->left;
        p = extract_messageid(overview->data);
        if (p == NULL) {
            warn("cannot find message ID in %s", filename);
            free(filename);
            free(article);
            continue;
        }
        if (HISlookup(history, p, &artdata.arrived, NULL, &artdata.expires,
                      &artdata.token)) {
            if (!tdx_data_store(data, &artdata))
                warn("cannot store data for %s", filename);
        } else {
            warn("cannot find article %s in history", p);
        }
        free(p);
        free(filename);
        free(article);
    }
    vector_free(files);
    vector_free(extra);

    info.indexinode = data->indexinode;
    info.base = data->base;
    if (!tdx_index_rebuild_finish(index, entry, &info))
        die("cannot update group index for %s", group);
    if (!tdx_data_rebuild_finish(group))
        die("cannot finish rebuilding data for group %s", group);
    tdx_data_close(data);
    HISclose(history);
}


/*
**  Parse an article number or range, returning the low and high values in the
**  provided arguments.  Returns true if the number or range was parsed
**  successfully, false otherwise.  Allows such constructs as "-", "20-", or
**  "-50" and leaves the unspecified ends of the range set to zero.
**
**  This code is similar to code in nnrpd and possibly should move into libinn
**  as common code.
*/
static bool
parse_range(char *range, ARTNUM *low, ARTNUM *high)
{
    char *hyphen, *end;

    *low = 0;
    *high = 0;
    hyphen = strchr(range, '-');
    if (hyphen == NULL) {
        *low = strtoul(range, &end, 10);
        if (*low == 0 || *end != '\0')
            return false;
        *high = *low;
        return true;
    } else {
        *hyphen = '\0';
        if (*range != '\0') {
            *low = strtoul(range, &end, 10);
            if (*low == 0 || *end != '\0')
                return false;
        }
        if (hyphen[1] != '\0') {
            *high = strtoul(hyphen + 1, &end, 10);
            if (*high == 0 || *end != '\0')
                return false;
        }
        *hyphen = '-';
        return true;
    }
}


/*
**  Main routine.  Load inn.conf, parse the arguments, and dispatch to the
**  appropriate function.
*/
int
main(int argc, char *argv[])
{
    int option;
    char flag = 'y';
    char mode = '\0';
    const char *newsgroup = NULL;
    const char *path = NULL;
    ARTNUM artlow = 0;
    ARTNUM arthigh = 0;

    message_program_name = "tdx-util";

    if (!innconf_read(NULL))
        exit(1);

    /* Parse options. */
    opterr = 0;
    while ((option = getopt(argc, argv, "a:f:n:p:AFR:cgiOo")) != EOF) {
        switch (option) {
        case 'a':
            if (!parse_range(optarg, &artlow, &arthigh))
                die("invalid article number or range %s", optarg);
            break;
        case 'f':
            flag = optarg[0];
            break;
        case 'n':
            newsgroup = optarg;
            break;
        case 'p':
            innconf->pathoverview = xstrdup(optarg);
            break;
        case 'A':
            if (mode != '\0')
                die("only one mode option allowed");
            mode = 'A';
            break;
        case 'F':
            if (mode != '\0')
                die("only one mode option allowed");
            mode = 'F';
            break;
        case 'R':
            if (mode != '\0')
                die("only one mode option allowed");
            mode = 'R';
            path = optarg;
            break;
        case 'c':
            if (mode != '\0')
                die("only one mode option allowed");
            mode = 'c';
            break;
        case 'g':
            if (mode != '\0')
                die("only one mode option allowed");
            mode = 'g';
            break;
        case 'i':
            if (mode != '\0')
                die("only one mode option allowed");
            mode = 'i';
            break;
        case 'O':
            if (mode != '\0')
                die("only one mode option allowed");
            mode = 'O';
            break;
        case 'o':
            if (mode != '\0')
                die("only one mode option allowed");
            mode = 'o';
            break;
        default:
            die("invalid option %c", optopt);
            break;
        }
    }

    /* Some modes require a group be specified. */
    if (strchr("cgoOR", mode) != NULL && newsgroup == NULL)
        die("group must be specified for -%c", mode);

    /* Run the specified function. */
    switch (mode) {
    case 'A':
        tdx_index_audit(false);
        break;
    case 'F':
        if (getenv("INN_TESTSUITE") == NULL)
            ensure_news_user_grp(true, true);
        tdx_index_audit(true);
        break;
    case 'R':
        if (getenv("INN_TESTSUITE") == NULL)
            ensure_news_user_grp(true, true);
        group_rebuild(newsgroup, path);
        break;
    case 'c':
        if (getenv("INN_TESTSUITE") == NULL)
            ensure_news_user_grp(true, true);
        group_create(newsgroup, artlow, arthigh, flag);
        break;
    case 'i':
        dump_index(newsgroup);
        break;
    case 'g':
        dump_group_index(newsgroup);
        break;
    case 'O':
        dump_overview(newsgroup, artlow, arthigh, true);
        break;
    case 'o':
        dump_overview(newsgroup, artlow, arthigh, false);
        break;
    default:
        die("a mode option must be specified");
        break;
    }
    exit(0);
}
